File: src/main/etl_job.py
from src.utils.spark_session import create_spark_session

def main():
    spark = create_spark_session("ETLExample")

    # Extraction: Read CSV
    df = spark.read.option("header", True).csv("data/input/sample_data.csv")

    # Cleans the data by removing any rows that contain null (missing) values and then removing duplicate rows.
    # df_clean will contain only unique and complete records with no missing data.
    # Transformation: cleaning and manipulating data
    df_clean = df.na.drop().dropDuplicates()

    # saves the cleaned data as a Parquet file in the folder data/output/cleaned_data.parquet.
    # The option "overwrite" means it will replace any existing file in that location.
    # Load: save result as Parquet
    df_clean.write.mode("overwrite").parquet("data/output/cleaned_data.parquet")

    print("✅ ETL completed successfully.!")
    spark.stop()

if __name__ == "__main__":
    main()

File: src/main/scala/com/projecticaro/etl/ETLJob.scala
package com.projecticaro.etl

import org.apache.spark.sql.SparkSession

object ETLJob {
  def main(args: Array[String]): Unit = {
    // Here Spark is initialized.
    // It creates the distributed engine that will execute your transformations and actions in parallel.
    // The location[*] indicates that it will use all the cores of your computer.
    val spark = SparkSession.builder()
      .appName("ETLProjectIcaroScala")
      .master("local[*]")
      .config("spark.sql.shuffle.partitions", "2")
      .getOrCreate()

    // Extraction: read CSV
    val df = spark.read.option("header", true).csv("data/input/sample_data.csv")

    // Transformation: remove nulls and duplicates
    val dfClean = df.na.drop().dropDuplicates()

    // Load: save to Parquet
    dfClean.write.mode("overwrite").parquet("data/output/cleaned_data_scala.parquet")

    println("✅ ETL completed successfully (Scala)!")
    spark.stop()
  }
}
